File: leetcode/159.longest-substring-with-at-most-two-distinct-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstringTwoDistinct(string s) {
        
        unordered_map<char, int> count_map;
        int j = 0;
        int res = 0;
        int k = 2;
        
        for (int i = 0; i < s.size(); ++i)
        {
            while(j < s.size() && (count_map.size() < k || (count_map.count(s[j]) > 0 && count_map.size() == k)))
            {
                count_map[s[j]] += 1;
                j ++;
            }
            
            res = max(res, j - i);
         
            if (count_map.count(s[i]) > 0){
                count_map[s[i]] -= 1;
                if (count_map[s[i]] == 0)
                {
                    count_map.erase(s[i]);
                }
            }
        }
        
        return res;
    }
};

File: leetcode/159.longest-substring-with-at-most-two-distinct-characters.py
class Solution(object):
    def lengthOfLongestSubstringTwoDistinct(self, s):
        """
        :type s: str
        :rtype: int
        """
        res = 0
        j = 0
        k = 2
        count_map = dict()

        for i in range(len(s)):
            while (j < len(s) and (len(count_map) < k or (len(count_map) == k and s[j] in count_map))):
                count_map[s[j]] = count_map.get(s[j], 0) + 1
                j += 1

            res = max(res, j - i)

            if s[i] in count_map:
                count_map[s[i]] -= 1
                if count_map[s[i]] == 0:
                    del count_map[s[i]]

        return res      

